### PR TITLE
Fixed #30981 -- Fixed admin changelist crash when using F() or OrderBy() expressions in admin_order_field.

### PR DESCRIPTION
**Pull request by hramezani**
[Ticket 30981](https://code.djangoproject.com/ticket/30981)

### CODE REVIEW

File: django/contrib/admin/views/main.py [hunk 1]
```diff
@@ -308,7 +308,9 @@ def get_ordering(self, request, queryset):
                     order_field = self.get_ordering_field(field_name)
                     if not order_field:
                         continue  # No 'admin_order_field', skip it
-                    if hasattr(order_field, 'as_sql'):
+                    if isinstance(order_field, OrderBy):
+                        ordering.append(order_field)
```
Comment (felixxm) [inline django/contrib/admin/views/main.py]: `OrderBy` should also support reverse ordering.

File: tests/admin_views/tests.py [hunk 2]
```diff
@@ -368,15 +368,25 @@ def test_change_list_sorting_callable_query_expression(self):
...
-            'Results of sorting on callable are out of order.'
-        )
-        self.assertContentBefore(
-            response, 'Middle content', 'Newest content',
-            'Results of sorting on callable are out of order.'
-        )
+        tests = (  # (description, index)
+            ('F("date") + DurationValue(3 days, 0:00:00)', 9),
+            ('F("date")', 12),
+            ('F("date").asc(nulls_last=True)', 13),
```
Comment (felixxm) [inline tests/admin_views/tests.py]: I think that property name i.e. `order_by_expression`, `order_by_f_expression`, or `order_by_orderby_expression` is a sufficient description.

File: tests/admin_views/tests.py [hunk 3]
```diff
@@ -368,15 +368,25 @@ def test_change_list_sorting_callable_query_expression(self):
...
+                self.assertContentBefore(
+                    response, 'Oldest content', 'Middle content',
+                    'Results of sorting on callable are out of order.'
+                )
+                self.assertContentBefore(
+                    response, 'Middle content', 'Newest content',
+                    'Results of sorting on callable are out of order.'
+                )
 
     def test_change_list_sorting_callable_query_expression_reverse(self):
```
Comment (felixxm) [inline tests/admin_views/tests.py]: We should add `-12` and `-13` to this test.

**Comment by hramezani** [comment]
@felixxm , Thanks for the review.
I changed the patch to support reverse ordering and also added a test for it.

Would it be good to add `toggle` to `django.db.models.expressions.OrderBy` :

```
    def toggle(self):
        self.descending = not self.descending
```

and use it to toggle the ordering. like: 
```
                    if isinstance(order_field, OrderBy):
                        if pfx == '-':  # Toggle ordering
                            order_field.toggle()
                        ordering.append(order_field)
```

File: django/contrib/admin/views/main.py (hunk 1 above)
Comment (felixxm) [reply django/contrib/admin/views/main.py]: We already have a  `reverse_ordering()`. Moreover this solution mutates `OrderBy` expression, so maybe:
```python
if isinstance(order_field, OrderBy):
    if pfx == '-':
        order_field = order_field.copy()
        order_field.reverse_ordering()
    ordering.append(order_field)
```

Comment (hramezani) [reply django/contrib/admin/views/main.py]: Thanks for the description. I replaced your suggested code block.

**Comment by felixxm** [comment]
@hramezani Thanks :+1: I pushed minor edits.

**Pull request merged**